src/services/mapCode.ts

```diff
@@ -0,0 +1,480 @@
+import {
+    Block,
+    ClassElement,
+    ClassLikeDeclaration,
+    createSourceFile,
+    factory,
+    FileTextChanges,
+    findAncestor,
+    flatten,
+    forEach,
+    ForInStatement,
+    formatting,
+    ForStatement,
+    getTokenAtPosition,
+    IfStatement,
+    InterfaceDeclaration,
+    isBlock,
+    isClassElement,
+    isClassLike,
+    isForInOrOfStatement,
+    isForStatement,
+    isIfStatement,
+    isInterfaceDeclaration,
+    isLabeledStatement,
+    isNamedDeclaration,
+    isSourceFile,
+    isTypeElement,
+    isWhileStatement,
+    LabeledStatement,
+    LanguageServiceHost,
+    Mutable,
+    Node,
+    NodeArray,
+    Program,
+    ScriptTarget,
+    SourceFile,
+    SourceMapper,
+    Statement,
+    SyntaxKind,
+    textChanges,
+    TextSpan,
+    TypeElement,
+    UserPreferences,
+    WhileStatement,
+} from "./_namespaces/ts";
+import { ChangeTracker } from "./textChanges";
+
+/** @internal */
+export function mapCode(
+    mappings: {
+        sourceFile?: SourceFile;
+        contents: string[];
+        focusLocations?: TextSpan[][];
+    }[],
+    _program: Program,
+    host: LanguageServiceHost,
+    formatContext: formatting.FormatContext,
+    preferences: UserPreferences,
+    _sourceMapper: SourceMapper,
+    _updates?: FileTextChanges[],
+): FileTextChanges[] {
+    try {
+        return textChanges.ChangeTracker.with(
+            { host, formatContext, preferences },
+            changeTracker => {
+                mappings.forEach(({ sourceFile, contents, focusLocations }) => {
+                    if (!sourceFile) {
+                        // TODO(@zkat): uhhh... do something about file-less mappings. Not supported for now.
+                        return;

```

Comment:
Same comment as before: Should we make this a `continue`?

Reply:
ditto: it's a `forEach`?